llvm/test/CodeGen/BPF/cttz-ctlz.ll

```diff
@@ -0,0 +1,304 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc < %s -march=bpf | FileCheck %s
+
+; test that we can expand CTTZ & CTLZ
+
+declare i32 @llvm.cttz.i32(i32, i1)
+
+define i32 @cttz_i32_zdef(i32 %a) {
+; CHECK-LABEL: cttz_i32_zdef:
```

Comment:
Question, how stable are these expansions?
Previously compiler would just error out, maybe just insert some dummy checks that verify that something is returned from these functions? And we can add a few tests on kernel side that verify runtime result. Wdyt?

Reply:
> Question, how stable are these expansions?

It is expanded by common codegen functions, thus might be changed without modifications in BPF backend.

> Previously compiler would just error out, maybe just insert some dummy checks that verify that something is returned from these functions? 

I think for compilers "CodeGen" tests, it is common to assert "Generated Asm" without actually run it. For example, for aarch64 backend developers, a "trick" is to write some code on x86 machine, testing the assembly without aarch64 emulator(e.g. qemu-user). 

So IMHO, for LLVM codegen we can just `CHECK-NEXT`: these asm lines. (They can be automatically generated)

> And we can add a few tests on kernel side that verify runtime result.

Yes, that would be nice if we can test "runtime" behavior.

Reply:
> I think for compilers "CodeGen" tests, it is common to assert "Generated Asm" without actually run it. For example, for aarch64 backend developers, a "trick" is to write some code on x86 machine, testing the assembly without aarch64 emulator(e.g. qemu-user).

Actually, I meant to skip matching the full expansion and just do something like below:

```
; CHECK-LABEL: cttz_i32_zdef:
; CHECK:         r0 =
; CHECK:         exit
```

(As we don't really test the expansion logic, we just test that some expansion is applied).

> So IMHO, for LLVM codegen we can just CHECK-NEXT: these asm lines. (They can be automatically generated)

You mean `utils/update_llc_test_checks`, right?
Well, maybe that's good enough, at-least the tests are easy to adjust if anything changes.

Reply:
> (As we don't really test the expansion logic, we just test that some expansion is applied)

Hmm, I'm worried about if we do at tests like

```
; CHECK-LABEL: cttz_i32_zdef:
; CHECK:         r0 =
; CHECK:         exit
```

how can this catch BPF backend changes like `setOperationAction()` to `Custom` and do some custom lowering, as if custom lowering might be incorrect though?

I think we should test implementation details about "how to expand / custom" actually. For example if someone really changes the logic of expansion, I think these tests shall be regenerated and we can see such patches (touched our backend tests)

Reply:
Ok, fair enough.